### ISSUE TITLE
Upgrade to VS2015

### ISSUE THREAD

**Issue by Nucleoprotein** [x360ce/x360ce#468]
I think we need to upgrade to VS2015. 
Anything blocking from App side @EJocys ?

**Comment by EJocys** [comment]
Agree. I use Visual Studio 2015 for long time already and nothing is blocking me when it comes to Application.

**Comment by Nucleoprotein** [comment]
Done in 5a757d29e0348e738fec2becff00815ee4a553ce 
@EJocys please check if anything is broken.
@Squall-Leonhart Can you check some games that have problems with VS2013 builds but works fine with VS2010 now with VS2015 Update 3 + KB3165756?

**Comment by EJocys** [comment]
@Nucleoprotein Looks like MiniHook was moved and not added to the source control:

Severity    Code    Description Project File    Line    Source  Suppression State
Error   C1083   Cannot open source file: '....\MinHook\src\buffer.c': No such file or directory    InputHook   D:\Projects\TocaEdit\x360ce\InputHook\c1    1   Build   

P.S.: Sometimes, I use this script to get clean and fresh copy from source control to see if it builds without the issues:

:: TocaEdit_clone_as_SVN.bat
::-------------------------------------------------------------
SET prg="%ProgramFiles%\TortoiseSVN\bin\svn.exe"
IF NOT EXIST %prg% SET prg="%ProgramFiles(x86)%\TortoiseSVN\bin\svn.exe"
IF NOT EXIST %prg% SET prg="%ProgramW6432%\TortoiseSVN\bin\svn.exe"
%prg% checkout https://github.com/x360ce/x360ce.git/trunk ".\TocaEdit"
%prg% checkout https://github.com/TsudaKageyu/minhook.git/trunk ".\TocaEdit\MinHook"
pause

**Comment by Nucleoprotein** [comment]
You need to initalize and update git submodules.

EDIT: Dunno if SVN can do this. MinHook is now submodule and is downloaded from TsudaKageyu/minhook github repository.

**Comment by EJocys** [comment]
I use SVN and it looks like GitHub doesn't support "svn:externals".
https://help.github.com/articles/subversion-properties-supported-by-github/
Added MinHook as a normal checkout.

---

Building original "libMinHook" inside original solution D:\Projects\TocaEdit\MinHook\build\VC15\MinHookVC15.sln produces no warnings.

Building "InputHook" reports these warnings:

2>d:\projects\tocaedit\minhook\src\trampoline.c(281): warning C4244: '+=': conversion from 'UINT' to 'UINT8', possible loss of data
2>  hook.c
2>d:\projects\tocaedit\minhook\src\hook.c(70): warning C4214: nonstandard extension used: bit field types other than int
2>d:\projects\tocaedit\minhook\src\hook.c(71): warning C4214: nonstandard extension used: bit field types other than int
2>d:\projects\tocaedit\minhook\src\hook.c(72): warning C4214: nonstandard extension used: bit field types other than int
2>d:\projects\tocaedit\minhook\src\hook.c(379): warning C4310: cast truncates constant value
2>d:\projects\tocaedit\minhook\src\hook.c(395): warning C4244: '=': conversion from 'BOOL' to 'UINT8', possible loss of data
2>d:\projects\tocaedit\minhook\src\hook.c(396): warning C4244: '=': conversion from 'BOOL' to 'UINT8', possible loss of data
2>d:\projects\tocaedit\minhook\src\hook.c(570): warning C4244: '=': conversion from 'BOOL' to 'UINT8', possible loss of data
2>d:\projects\tocaedit\minhook\src\hook.c(750): warning C4244: '=': conversion from 'BOOL' to 'UINT8', possible loss of data
2>d:\projects\tocaedit\minhook\src\hook.c(757): warning C4244: '=': conversion from 'BOOL' to 'UINT8', possible loss of data
2>d:\projects\tocaedit\minhook\src\hook.c(847): warning C4054: 'type cast': from function pointer 'FARPROC' to data pointer 'LPVOID'
2>  hde32.c
2>  buffer.c
2>d:\projects\tocaedit\minhook\src\buffer.c(52): warning C4201: nonstandard extension used: nameless struct/union
2>d:\projects\tocaedit\minhook\src\buffer.c(154): warning C4100: 'pOrigin': unreferenced formal parameter
2>  Generating Code...
2>d:\projects\tocaedit\minhook\src\hde\hde32.c(60): warning C4701: potentially uninitialized local variable 'c' used
2>d:\projects\tocaedit\minhook\src\hook.c(248): warning C4701: potentially uninitialized local variable 'enable' used

**Comment by Nucleoprotein** [comment]
Please use git instead of svn, I will change versioning system in this week to use git. Git is more powerful than SVN and I really like it, maybe is little hard to learn at start but is much better when you learn how to use it.
I know about warning, ignore them for now.
I think you can clone repository with submodules using `git clone --recursive`

**Comment by EJocys** [comment]
> Git is more powerful than SVN and I really like it

Agree, but when it comes to usability it is pain in the ass and complete mess. GIT UI developers have not figured out that it needs "simple mode":

```
a) hiding "stash" feature and
b) auto-commit to remote repository (i.e hiding local repository).

http://obaida.info/blog/2015/03/git-github/
```

At current state, TortoiseGIT just can't compete with convenience of TortoiseSVN and VisualSVN. Visual Studio integration of GIT is horrible when compared to VisualSVN.

I will switch to GIT when Windows Explorer and Visual Studio integration will be as good as SVN integration. At the moment, VisualSVN features counters inconvenience caused by GIT not supporting svn:externals

**Comment by Nucleoprotein** [comment]
Usability of git is greater than SVN if you read it's small, pocket 16 ton manual :)
TortoiseGIT ? Is very complete, you can make all possible operation using it. MS VS2015 integration is a little pain, but I do not use it at all, TortoiseGIT is sufficient - yesterday I downloaded PR of Dolphin and rebased in over master without any problems.
I think main problem is different idea of how repositories works in git, because they work in different way than in SVN.

NOTE: I found 

```
:: TocaEdit_clone_as_GIT.bat
::-------------------------------------------------------------
SET prg="%ProgramFiles%\Git\bin\git.exe"
IF NOT EXIST %prg% SET prg="%ProgramFiles(x86)%\Git\bin\git.exe"
IF NOT EXIST %prg% SET prg="%ProgramW6432%\Git\bin\git.exe"
%prg% clone https://github.com/x360ce/x360ce.git ".\TocaEdit"
pause
```

which is incorrect, you should never use git.exe from bin, use git.exe from C:\Program Files\Git\cmd.

**Comment by EJocys** [comment]
> you should never use git.exe from bin, use git.exe from C:\Program Files\Git\cmd.

Thanks. Updated.

> I think main problem is different idea of how repositories works in git

One of the main problems of understanding GIT is, that it litters menus with non-required options and wording is a mess. Properly designed project would match antonyms to related synonyms. For example if developer decides to use "check-out" word then opposite action must use "check-in" word. Same goes for "pull" and "push". Looking at GIT it is clear that most of the effort went into features and no effort spent on interface or proper wording. It will be much harder to fix that now when GIT is used worldwide.

TortoiseGIT could use some logic inside menus like:

-- Simple and single action menus:
GIT Commit and Push...
GIT Pull and Update <- single click update
-- Optional menus for advanced users:
GIT/Commit to Local Repository...
GIT/Update from Local Repository...
GIT/Local Repository/Push to Remote Repository...
GIT/Local Repository/Pull from Remote Repository...

Instead of some vague and telling nothing "Commit.." or "Push...". On TortoiseSVN you ca update project with ONE mouse click on the menu. I can't understand why there is no such menu option on TortoiseGIT when this is the most useful feature and ca be easily implemented. Maybe I need to ask TortoiseGIT devs :).

Local repository is an optional feature for small teams and simple projects. In my, opinion, best practice is to split big changes into smaller parts and make sure that project will work after each small part is applied and committed to remote repository. Local repository encourages, to pile multiple changes locally and this can lead to conflicts later (when other developers change same parts).

I am sure GIT will get there, but it will be hard to beat SVN which offers exactly what small team needs without GIT learning complexity.

**Comment by Nucleoprotein** [comment]
Hym, you use latest TortoiseGit version ? I always use TortoiseGit Sync menu, it's very consistent - you have there all commands to synchronize your repository.

About local repositories... i think this is mian git advantage ! You can easily create all patches in branches and add one by one to master by using pull requests ! You only need to rebase if master was changed to much, but in most cases it as simple as 3 clicks. Github can also merge PR automatically :)
You can also add multiple origins to your git repository etc. (this is useful for rebasing PR locally, i used this for Dolphin PR then rebased it over master).

**Comment by EJocys** [comment]
> I always use "TortoiseGit Sync..." menu, it's very consistent

"TortoiseGit Sync..." opens new Window with bunch of buttons, while "SVN Update" menu does full update from remote repository immediately which is extremely useful and simple. I use this button 5-10 times a day at work, due to the frequent updates to the source control - immediately after SVN monitor notifies me about changes done by other developers. This reduce conflicts to 0, because they are resolved on the fly.

> About local repositories... i think this is main git advantage ! 

You are right. But this advantage comes at the price - complexity. If this advantage was optional (hidden and maintained automatically) then SVN would be dead years ago. GIT have interface usability issue, like Linux/Unix/BSD have against Windows or Apple. Most people don't care how many features it can offer if they don't need these features, especially if these features cluttering interface. As, I've said, I prefer to do small changes and commit them frequently to the root repository, because it helps tremendously to solve conflicts while they are still small. This strategy makes local repository redundant, because there is no need to have repository for drafts. I am not against local repository, I am just saying that 1/3 of developers have no practical use for it and GIT must address it by making visibility of local repository optional, otherwise GIT will never completely replace SVN.

**Comment by EvoJocys** [comment]
Done in the 4.x line.

Closing as part of a tracker clean-up after the 4.21 release.
